rules/namespaces.go

```diff
@@ -0,0 +1,160 @@
+// Copyright (c) 2017 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package rules
+
+import (
+	"sync"
+	"time"
+
+	"github.com/m3db/m3cluster/kv"
+	"github.com/m3db/m3metrics/generated/proto/schema"
+	"github.com/m3db/m3metrics/policy"
+	"github.com/m3db/m3x/errors"
+	"github.com/m3db/m3x/log"
+)
+
+var emptyNamespaces policy.Namespaces
+
+// namespaces contains the list of namespace users have defined
+// rules for. The namespace list is updated as needed.
+type namespaces struct {
+	sync.RWMutex
+	*runtimeValue
+
+	key          string
+	store        kv.Store
+	opts         Options
+	ruleSetKeyFn RuleSetKeyFn
+	log          xlog.Logger
+
+	proto *schema.Namespaces
+	rules map[string]*ruleSets
+}
+
+func newNamespaces(key string, store kv.Store, opts Options) *namespaces {
+	n := &namespaces{
+		key:          key,
+		store:        store,
+		opts:         opts,
+		ruleSetKeyFn: opts.RuleSetKeyFn(),
+		log:          opts.InstrumentOptions().Logger(),
+		proto:        &schema.Namespaces{},
+		rules:        make(map[string]*ruleSets),
+	}
+	n.runtimeValue = newRuntimeValue(key, store, n.toNamespaces, namespacesVersion, namespacesCutover, n.process, opts)
```

Comment:
I'd prefer the users to explicitly call `Start`, which makes it more clear when the values start updating, and facilitates testing as well.